From ffd445717fb4e4d9509ee21d3dbdc76213f6b211 Mon Sep 17 00:00:00 2001
From: Ikey Doherty <ikey@solus-project.com>
Date: Sun, 16 Apr 2017 06:09:18 +0100
Subject: [PATCH] panel: Fix troublesome popovers once and for all.

This commit fixes two issues: GTK 3.22 animations (doubling) and, finally,
the glitchy menu on the bottom panel. Granted, the code is especially evil,
but it actually is enough to trick GTK+ past it's own trip-ups and place
the popover in the right position every time.

This fixes 292 and it fixes 809 properly too.

Signed-off-by: Ikey Doherty <ikey@solus-project.com>
---
 src/panel/panel.vala    | 10 ----------
 src/panel/popovers.vala | 26 +++++++++++++++++++++++---
 2 files changed, 23 insertions(+), 13 deletions(-)

diff --git a/src/panel/panel.vala b/src/panel/panel.vala
index 500defe..a0a6851 100644
--- a/src/panel/panel.vala
+++ b/src/panel/panel.vala
@@ -998,16 +998,6 @@ public class Panel : Budgie.Toplevel
         while (Gtk.events_pending()) {
             Gtk.main_iteration();
         }
-
-        if (expanded) {
-            Idle.add(()=> {
-                if (get_window() != null) {
-                    get_window().focus(Gdk.CURRENT_TIME);
-                }
-                present();
-                return false;
-            });
-        }
     }
 
     void placement()
diff --git a/src/panel/popovers.vala b/src/panel/popovers.vala
index 1245b18..1bab779 100644
--- a/src/panel/popovers.vala
+++ b/src/panel/popovers.vala
@@ -33,7 +33,7 @@ public class PopoverManagerImpl : PopoverManager, GLib.Object
                 return Gdk.EVENT_PROPAGATE;
             }
             if (visible_popover != null) {
-                visible_popover.hide();
+                visible_popover.popdown();
                 make_modal(visible_popover, false);
                 visible_popover = null;
             }
@@ -54,7 +54,7 @@ public class PopoverManagerImpl : PopoverManager, GLib.Object
             }
             if ((e.x < alloc.x || e.x > alloc.x+alloc.width) ||
                 (e.y < alloc.y || e.y > alloc.y+alloc.height)) {
-                    visible_popover.hide();
+                    visible_popover.popdown();
                     make_modal(visible_popover, false);
                     visible_popover = null;
             }
@@ -101,8 +101,28 @@ public class PopoverManagerImpl : PopoverManager, GLib.Object
             return;
         }
 
+        pop.set_modal(false);
         owner.set_expanded(true);
-        pop.show();
+        Idle.add(()=> {
+            Gtk.Widget? relative_to = pop.get_relative_to();
+            if (owner.get_window() != null) {
+                owner.get_window().focus(Gdk.CURRENT_TIME);
+            }
+            owner.present();
+            pop.set_relative_to(null);
+            pop.set_relative_to(relative_to);
+            pop.queue_resize();
+            while (Gtk.events_pending()) {
+                Gtk.main_iteration();
+            }
+            Idle.add(()=>{
+                if (!pop.get_visible()) {
+                    pop.popup();
+                }
+                return false;
+            });
+            return false;
+        });
     }
 
     void on_widget_mapped(Gtk.Widget? p)
-- 
2.11.0

